### PR TITLE
Iss hipcms 477 angular e2e protractor testing

### PR DESCRIPTION
**Pull request by PoornimaKaruppasamy**
Sample e2e tests written for our application - For reference: https://atlassian-hip.cs.uni-paderborn.de/confluence/display/DCS/Server-side+and+client-side+concept+for+unit+testing

### CODE REVIEW

File: app/authentication/login/login.component.e2e-spec.js [hunk 1]
```diff
@@ -0,0 +1,31 @@
+"use strict";
```
Comment (venkateshpenchala) [inline app/authentication/login/login.component.e2e-spec.js]: __[jshint]__
* Use the function form of "use strict".

File: app/authentication/login/login.component.e2e-spec.js [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+var protractor_1 = require("protractor");
```
Comment (venkateshpenchala) [inline app/authentication/login/login.component.e2e-spec.js]: __[jshint]__
* 'require' is not defined.

File: app/authentication/login/login.component.e2e-spec.js [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+describe('Login', function () {
```
Comment (venkateshpenchala) [inline app/authentication/login/login.component.e2e-spec.js]: __[jshint]__
* 'describe' is not defined.

File: app/authentication/login/login.component.e2e-spec.js [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+    beforeEach(function () {
```
Comment (venkateshpenchala) [inline app/authentication/login/login.component.e2e-spec.js]: __[jshint]__
* 'beforeEach' is not defined.

File: app/authentication/login/login.component.e2e-spec.js [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+        protractor_1.browser.get('/login');
+    });
+    it('should have an input and submit button', function () {
```
Comment (venkateshpenchala) [inline app/authentication/login/login.component.e2e-spec.js]: __[jshint]__
* 'it' is not defined.

File: app/authentication/login/login.component.e2e-spec.js [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+        var allInputElements = protractor_1.element.all(protractor_1.by.css('md-card md-card-content form input'));
+        var inputElement = protractor_1.element(protractor_1.by.css('md-card md-card-content form input'));
+        var submitButton = protractor_1.element(protractor_1.by.css('md-card md-card-content form button'));
+        var loginCardTitle = protractor_1.element(protractor_1.by.css('md-card md-card-title h3'));
+        protractor_1.browser.wait(function () {
+            return inputElement.isPresent();
+        }, 60000);
+        expect(inputElement.isPresent()).toEqual(true);
```
Comment (venkateshpenchala) [inline app/authentication/login/login.component.e2e-spec.js]: __[jshint]__
* 'expect' is not defined.

File: app/authentication/login/login.component.e2e-spec.js [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+        expect(submitButton.isPresent()).toEqual(true);
```
Comment (venkateshpenchala) [inline app/authentication/login/login.component.e2e-spec.js]: __[jshint]__
* 'expect' is not defined.

File: app/authentication/login/login.component.e2e-spec.js [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+        expect(protractor_1.element(protractor_1.by.css('md-card md-card-title h3')).getText()).toEqual('Geben Sie Ihre Daten ein, um sich einzuloggen');
```
Comment (venkateshpenchala) [inline app/authentication/login/login.component.e2e-spec.js]: __[jshint]__
* 'expect' is not defined.

File: app/dashboard/dashboard.component.e2e-spec.js [hunk 9]
```diff
@@ -0,0 +1,15 @@
+"use strict";
```
Comment (venkateshpenchala) [inline app/dashboard/dashboard.component.e2e-spec.js]: __[jshint]__
* Use the function form of "use strict".

File: app/dashboard/dashboard.component.e2e-spec.js [hunk 10]
```diff
@@ -0,0 +1,15 @@
...
+var protractor_1 = require("protractor");
+describe('Dashboard', function () {
```
Comment (venkateshpenchala) [inline app/dashboard/dashboard.component.e2e-spec.js]: __[jshint]__
* 'describe' is not defined.

File: app/dashboard/dashboard.component.e2e-spec.js [hunk 11]
```diff
@@ -0,0 +1,15 @@
...
+    beforeEach(function () {
+        protractor_1.browser.get('/dashboard');
+    });
+    it('should show dashboard', function () {
+        protractor_1.browser.wait(function () {
+            return protractor_1.element(protractor_1.by.css('hip-dashboard md-card h2')).isPresent();
+        }, 60000);
+        expect(protractor_1.element(protractor_1.by.css('hip-dashboard md-card h2')).isPresent()).toEqual(true);
```
Comment (venkateshpenchala) [inline app/dashboard/dashboard.component.e2e-spec.js]: __[jshint]__
* 'expect' is not defined.

File: app/dashboard/dashboard.component.e2e-spec.js [hunk 12]
```diff
@@ -0,0 +1,15 @@
...
+        expect(protractor_1.element.all(protractor_1.by.css('hip-dashboard md-card h2')).get(0).getText()).toEqual('Dashboard');
```
Comment (venkateshpenchala) [inline app/dashboard/dashboard.component.e2e-spec.js]: __[jshint]__
* 'expect' is not defined.

File: protractor.config.js [hunk 13]
```diff
@@ -34,7 +34,7 @@ exports.config = {
   useAllAngular2AppRoots: true,
 
   // Base URL for application server
-  baseUrl: 'http://localhost:8080',
+  baseUrl: 'http://localhost:3000',
```
Comment (maltehol) [inline protractor.config.js]: We do not need to start a lite-server with browser-sync, which will be started with npm start and run on port 3000. It is enough to have a simple http server you can start with  npm run serve which will run on port 8080.

File: app/authentication/login/login.component.e2e-spec.ts [hunk 14]
```diff
@@ -0,0 +1,40 @@
...
+      return inputElement.isPresent();
+    }, 60000);
+
+    expect(inputElement.isPresent()).toEqual(true);
+    expect(submitButton.isPresent()).toEqual(true);
+    expect(element(by.css('md-card md-card-title h3')).getText()).toEqual('Geben Sie Ihre Daten ein, um sich einzuloggen');
+
+    //send inputs
+    allInputElements.first().sendKeys('admin@hipapp.de');
+    allInputElements.last().sendKeys('Hipcms@123');
```
Comment (maltehol) [inline app/authentication/login/login.component.e2e-spec.ts]: Even in Test code: do not hardcode credentials.
We are working on an *open source* project. So everyone can look in our code and have now our login data to login as admin!

Better:
first create a user on the sign up page and then login with these credentials of the newly created user.

File: app/dashboard/dashboard.component.e2e-spec.js [hunk 15]
```diff
@@ -0,0 +1,15 @@
...
+    });
+});
+//# sourceMappingURL=dashboard.component.e2e-spec.js.map
```
Comment (maltehol) [inline app/dashboard/dashboard.component.e2e-spec.js]: add e2e-spec.js files to .gitignore

File: app/authentication/login/login.component.e2e-spec.ts [hunk 16]
```diff
@@ -0,0 +1,40 @@
...
+    let inputElement = element(by.css('md-card md-card-content form input'));
+    let submitButton = element(by.css('md-card md-card-content form button'));
+    let loginCardTitle = element(by.css('md-card md-card-title h3'));
+
+    browser.wait(function () {
+      return inputElement.isPresent();
+    }, 60000);
+
+    expect(inputElement.isPresent()).toEqual(true);
+    expect(submitButton.isPresent()).toEqual(true);
```
Comment (maltehol) [inline app/authentication/login/login.component.e2e-spec.ts]: The `toEqual(true)` is redudant

File: app/authentication/login/login.component.e2e-spec.ts [hunk 17]
```diff
@@ -0,0 +1,40 @@
...
+    expect(element(by.css('md-card md-card-title h3')).getText()).toEqual('Geben Sie Ihre Daten ein, um sich einzuloggen');
```
Comment (maltehol) [inline app/authentication/login/login.component.e2e-spec.ts]: the `h3` is in the current code a `p`. But it's enough to select the element by `md-card-title`

File: app/authentication/login/login.component.e2e-spec.ts [hunk 18]
```diff
@@ -0,0 +1,40 @@
...
+  beforeEach(() => {
+    browser.get('/login');
+  });
+
+  it('should have an input and submit button', () => {
+
+    let allInputElements = element.all(by.css('md-card md-card-content form input'));
+    let inputElement = element(by.css('md-card md-card-content form input'));
+    let submitButton = element(by.css('md-card md-card-content form button'));
+    let loginCardTitle = element(by.css('md-card md-card-title h3'));
```
Comment (maltehol) [inline app/authentication/login/login.component.e2e-spec.ts]: variable is not used

File: app/dashboard/dashboard.component.e2e-spec.ts [hunk 19]
```diff
@@ -0,0 +1,21 @@
...
+describe('Dashboard', () => {
+
+  beforeEach(() => {
+    browser.get('/dashboard');
+  });
+
+
+  it('should show dashboard', () => {
+    browser.wait(function () {
+      return element(by.css('hip-dashboard md-card h2')).isPresent();
```
Comment (maltehol) [inline app/dashboard/dashboard.component.e2e-spec.ts]: In the current code the title is an `h1` element.
But it's enough to select with `hip-dashboard md-card` or even `md-card`

File: app/dashboard/dashboard.component.e2e-spec.ts [hunk 20]
```diff
@@ -0,0 +1,21 @@
...
+     },60000);
+
+    expect(element(by.css('hip-dashboard md-card h2')).isPresent()).toEqual(true);
+    expect(element.all(by.css('hip-dashboard md-card h2')).get(0).getText()).toEqual('Dashboard' );
```
Comment (maltehol) [inline app/dashboard/dashboard.component.e2e-spec.ts]: you refer this element three times. You should create a variable for that to improve readability for your code

**Comment by maltehol** [comment]
please delete the generated `.js` files.

It seems, you did not understand, what the npm run e2e command does.

In confluence you wrote:

>  In first terminal -  npm start to run our application,
> In second terminal -  webdriver-manager start (to run tests in local browser),
> In third terminal - npm run serve (to make sure angular http server start)
>  In fourth terminal - npm run e2e (for protractor tests)

that's nonesense.
npm start, does a compilation of the files and also serve the files, while watching for changes.
npm run server, just creates a http server, which serves all the files.
npm run e2e dos multile things:
1. it compiles the application
2. it kills all other http-servers and create a new one, which serve files
3. it runs the test.

Please consider, that the tests will run on TFS, means that we have no monitor, where we can start a Browser. Also we can not open multiple terminals, so we should only have ONE command to do the testing (npm run e2e).
please research on headless testing for the issue with no monitor.

**Comment by PoornimaKaruppasamy** [comment]
I understood the way it runs. I discussed with Patrick in this morning regarding running e2e tests in TFS. Will make that as one command, which indeed not required to open multiple terminals.

As, in command npm run e2e , it is not written to listen on port 8080, in which npm run serve does.
Now, we have added that in npm run e2e. I will push the changes soon. We found like webdriver will start automatically on running the same. 

I will change in the confluence document and do research on headless testing.

File: app/dashboard/dashboard.component.e2e-spec.ts [hunk 21]
```diff
@@ -0,0 +1,22 @@
...
+    browser.wait(function () {
+      return dashboardCardElement.isPresent();
+    }, 60000);
+
+    expect(dashboardCardElement.getText()).toEqual('Dashboard');
+
+  });
+
+
+});
```
Comment (HIPCMSOrgaUser) [inline app/dashboard/dashboard.component.e2e-spec.ts]: __[tslint]__
* file should end with a newline

File: app/authentication/login/login.component.e2e-spec.ts [hunk 22]
```diff
@@ -0,0 +1,38 @@
...
+var testDataJson = require('../../../hip-config.json');
+
+describe('Login', () => {
+
+  beforeEach(() => {
+    browser.get('/login');
+  });
+
+  it('should have an input and submit button', () => {
+    //console.log(ConfigService);
```
Comment (HIPCMSOrgaUser) [inline app/authentication/login/login.component.e2e-spec.ts]: __[tslint]__
* comment must start with a space

File: app/authentication/login/login.component.e2e-spec.ts [hunk 23]
```diff
@@ -0,0 +1,38 @@
...
+    let loginCardTitle = element(by.css('md-card md-card-title'));
+
+    browser.wait(function () {
+      return inputElement.isPresent();
+    }, 60000);
+
+    expect(submitButton.isPresent()).toEqual(true);
+    expect(loginCardTitle.getText()).toEqual('Geben Sie Ihre Daten ein, um sich einzuloggen');
+
+    //send inputs
```
Comment (HIPCMSOrgaUser) [inline app/authentication/login/login.component.e2e-spec.ts]: __[tslint]__
* comment must start with a space

File: app/authentication/login/login.component.e2e-spec.ts [hunk 24]
```diff
@@ -0,0 +1,38 @@
...
+    allInputElements.first().sendKeys(testDataJson.userName);
+    allInputElements.last().sendKeys(testDataJson.passWord);
+
+    //click function & redirect
```
Comment (HIPCMSOrgaUser) [inline app/authentication/login/login.component.e2e-spec.ts]: __[tslint]__
* comment must start with a space

File: app/authentication/login/login.component.e2e-spec.ts [hunk 25]
```diff
@@ -0,0 +1,38 @@
...
+    submitButton.click().then(() => {
+      browser.driver.wait(function () {
+        return browser.driver.getCurrentUrl().then(function (url) {
+          return (/dashboard/).test(url);
+        });
+      }, 50000);
+    });
+
+  });
+});
```
Comment (HIPCMSOrgaUser) [inline app/authentication/login/login.component.e2e-spec.ts]: __[tslint]__
* file should end with a newline

File: app/authentication/login/login.component.e2e-spec.ts [hunk 26]
```diff
@@ -0,0 +1,38 @@
+import { browser, by, element } from 'protractor';
+var testDataJson = require('../../../hip-config.json');
```
Comment (HIPCMSOrgaUser) [inline app/authentication/login/login.component.e2e-spec.ts]: __[tslint]__
* forbidden var keyword

File: app/authentication/login/login.component.e2e-spec.ts [hunk 27]
```diff
@@ -0,0 +1,38 @@
...
+    passwordInput.sendKeys(testDataJson.password);
+
+    submitButton.click().then(() => {
+      browser.wait(function () {
+        return browser.getCurrentUrl().then(function (url) {
+          return (/dashboard/).test(url);
+        });
+      }, 50000);
+    });
+  })
```
Comment (HIPCMSOrgaUser) [inline app/authentication/login/login.component.e2e-spec.ts]: __[tslint]__
* missing semicolon

**Comment by maltehol** [comment]
TFS is setup, so it uses PhantomJs for now, instead of a installed browser. The reason is, the only browser installed on our TFS-Agent is Edge and protractor does not support edge yet.

The adaption for phantomjs in the configuration files is done in TFS so there is no need to change the configuration file in Github and the test can be run with an actual browser locally.

**Pull request merged**